nemo/core/optim/distributed_adam.py

```diff
@@ -278,13 +278,289 @@ def state_dict(self, *args, **kwargs):
+                raise ValueError(f'Param id {param_id} does not match any model sharded param') from e
+
+        # FP32 model params
+        optim_state_dict['fp32_optim_fp32_params'] = [
+            make_sharded_optimizer_tensor(get_safe(param_id), fp32_param, prefix=f'optimizer.state.fp32_from_fp16')
+            for param_id, fp32_param in enumerate(optim_state_dict['fp32_optim_fp32_params'])
+        ]
+
+        # FP32 Optimizer state
+        optim_state_to_sharding_state(fp32_state_dict, id_to_sharded_param_map, exclude_keys=('step',))
+
+        # Since this is a wrapped optimizer, we don't want to store hyperparameters
+        # but they must be updated with `update_fp32_hyperparameters` before calling `load_state_dict`
+        for group_idx in range(len(fp32_state_dict['param_groups'])):
+            # unwrap LocalNonpersitentObject from 'params' ...
+            fp32_state_dict['param_groups'][group_idx]['params'] = fp32_state_dict['param_groups'][group_idx][
+                'params'
+            ].obj
+            # ... and apply it to the whole group
+            fp32_state_dict['param_groups'][group_idx] = LocalNonpersitentObject(
+                fp32_state_dict['param_groups'][group_idx]
+            )
+
+        optim_state_dict['fp32_optim'] = fp32_state_dict
+
+    def update_fp32_hyperparameters(self, state_dict):
+        """ Copy relevant optimizer hyperparameters and step from main optimizer to FP32 one. """
+        if 'fp32_optim' not in state_dict:
+            return
+        for main_group, fp32_group in zip(state_dict['param_groups'], state_dict['fp32_optim']['param_groups']):
+            for k, v in main_group.items():
+                if k in fp32_group and k != 'params' and fp32_group[k] != v:
+                    logger.info(f'Replacing FP32 optimizer hparam {k} with {v} (previous value: {fp32_group[k]})')
+                    fp32_group[k] = v
+        # Copt step info
+        step = state_dict['state']['step']
+        for _, param_state in state_dict['fp32_optim']['state'].items():
+            param_state['step'] = step
+
+
+def adam_init_state(opt):
+    for group in opt.param_groups:
+        for p in group['params']:
+            if len(opt.state[p]) == 0:
+                opt.state[p]['exp_avg'] = torch.zeros_like(p.data)
+                opt.state[p]['exp_avg_sq'] = torch.zeros_like(p.data)
+
+
+def split_fp32(x):
+    assert x.dtype is torch.float, x.dtype
+    x = x.clone().detach()
+    rem_bf16 = x.unsqueeze(-1).view(torch.int16)
+    rem = rem_bf16[..., 0]
+    bf16 = rem_bf16[..., 1]
+    assert x.shape == rem.shape == bf16.shape, (x.shape, rem.shape, bf16.shape)
+    # Round up BF16
+    bf16 += torch.where(rem < 0, 1, 0)
+    return bf16.view(torch.bfloat16), rem
+
+
+def merge_fp32(bf16, rem):
+    assert bf16.dtype is torch.bfloat16, bf16.dtype
+    assert rem.dtype is torch.int16, rem.dtype
+    # Round down BF16
+    bf16 = bf16.clone().detach()
+    bf16 -= torch.where(rem < 0, 1, 0)
```

Comment:
I think we need to cast to int16 first before undoing the rounding:
```suggestion
    bf16 = bf16.clone().detach().view(torch.int16)
    bf16 -= torch.where(rem < 0, 1, 0)
```